packages/multi-test/src/app.rs

```diff
@@ -19,153 +19,345 @@ use crate::wasm::{ContractData, Wasm, WasmKeeper};
-        self.bank = Some(Box::new(bank));
-        self
+    pub fn with_bank<NewBank: Bank>(
+        self,
+        bank: NewBank,
+    ) -> AppBuilder<NewBank, ApiT, StorageT, CustomT, WasmT> {
+        let AppBuilder {
+            wasm,
+            api,
+            storage,
+            custom,
+            block,
+            ..
+        } = self;
+
+        AppBuilder {
+            wasm,
+            bank,
+            api,
+            storage,
+            custom,
+            block,
+        }
     }
 
     /// Overwrites default api interface
-    #[track_caller]
-    pub fn with_api(mut self, api: impl Api + 'static) -> Self {
-        assert!(self.api.is_none(), "API interface already overwritten");
-        self.api = Some(Box::new(api));
-        self
+    pub fn with_api<NewApi: Api>(
+        self,
+        api: NewApi,
+    ) -> AppBuilder<BankT, NewApi, StorageT, CustomT, WasmT> {
+        let AppBuilder {
+            wasm,
+            bank,
+            storage,
+            custom,
+            block,
+            ..
+        } = self;
+
+        AppBuilder {
+            wasm,
+            bank,
+            api,
+            storage,
+            custom,
+            block,
+        }
     }
 
     /// Overwrites default storage interface
-    #[track_caller]
-    pub fn with_storage(mut self, storage: impl Storage + 'static) -> Self {
-        assert!(
-            self.storage.is_none(),
-            "Storage interface already overwritten"
-        );
-        self.storage = Some(Box::new(storage));
-        self
+    pub fn with_storage<NewStorage: Storage>(
+        self,
+        storage: NewStorage,
+    ) -> AppBuilder<BankT, ApiT, NewStorage, CustomT, WasmT> {
+        let AppBuilder {
+            wasm,
+            api,
+            bank,
+            custom,
+            block,
+            ..
+        } = self;
+
+        AppBuilder {
+            wasm,
+            bank,
+            api,
+            storage,
+            custom,
+            block,
+        }
     }
 
     /// Overwrites default custom messages handler
-    #[track_caller]
-    pub fn with_custom(mut self, custom: impl CustomHandler<ExecC, QueryC> + 'static) -> Self {
-        assert!(self.custom.is_none(), "Custom handler already overwritten");
-        self.custom = Some(Box::new(custom));
-        self
+    ///
+    /// At this point it is needed that new custom implements some `CustomHandler` trait, but it doesn't need
+    /// to be bound to ExecC or QueryC yet - as those may change. The cross-components validation is
+    /// done on final building.
+    ///
+    /// Also it is possible to completely abandon trait bounding here which would not be bad idea,
+    /// however it might make the message on build creepy in many cases, so as for properly build
+    /// `App` we always want `Wasm` to be `Wasm`, some checks are done early.
+    pub fn with_custom<NewCustom: CustomHandler>(
```

Comment:
Same as above. There is an implicit connection between the ExecC/QueryC types in Query and Wasm and it would be great if the builder enforced they are compatible.

It seems this will then cause some issue later on when trying to use the resulting app that some method is not available (as impl requires a match). But ideally the compile error would appear on the line that caused the issue (this one when setting an incompatible Custom handler)

Reply:
And as before, it is in `build` by bound `WasmT: Wasm<CustomT::ExecC, CustomT::QueryC>`.